File: 234-palindrome-linked-list/palindrome-linked-list.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function isPalindrome(head: ListNode | null): boolean {
    if (!head || !head.next) return true

    let slow = head
    let fast = head
    let prev = null

    while (fast && fast.next) {
        fast = fast.next.next
        slow = slow.next
    }

    prev = null
    let next = null
    while (slow) {
        next = slow.next
        slow.next = prev
        prev = slow
        slow = next
    }

    let left = head
    let right = prev

    while (right) {
        if (left.val !== right.val) {
            return false
        }
        left = left.next
        right = right.next
    }

    return true
};

File: 234-palindrome-linked-list/palindrome-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        if not head or not head.next:
            return True

        # Phase 1 Find middle
        slow, fast = head, head
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next

        # Phase 2 Reverse second half
        prev = None
        curr = slow
        while curr:
            next_node = curr.next
            curr.next = prev
            prev = curr
            curr = next_node

        # prev is now head of the reversed second half

        # Phase 3 Compare the two halves
        left, right = head, prev
        while right:
            if left.val != right.val:
                return False
            left = left.next
            right = right.next

        return True